extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java

```diff
@@ -0,0 +1,94 @@
+
+import org.eclipse.dataspaceconnector.iam.did.spi.credentials.CredentialsVerifier;
+import org.eclipse.dataspaceconnector.iam.did.spi.document.DidDocument;
+import org.eclipse.dataspaceconnector.identityhub.client.IdentityHubClient;
+import org.eclipse.dataspaceconnector.identityhub.credentials.VerifiableCredentialsJwtUtils;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.result.AbstractResult;
+import org.eclipse.dataspaceconnector.spi.result.Result;
+
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Obtains and verifies credentials associated with a DID.
+ * The DID document contains an IdentityHub service, the IdentityHubCredentialsVerifier gets credentials from the
+ * IdentityHub instance and verifies the credentials.
+ */
+public class IdentityHubCredentialsVerifier implements CredentialsVerifier {
+
+    private static final String IDENTITY_HUB_SERVICE_TYPE = "IdentityHub";
+    private final IdentityHubClient identityHubClient;
+    private final Monitor monitor;
+    private final JwtCredentialsVerifier jwtCredentialsVerifier;
+    private final VerifiableCredentialsJwtUtils jwtPayloadParser;
+
+    /**
+     * Create a new credential verifier that uses an Identity Hub
+     *
+     * @param identityHubClient IdentityHubClient.
+     */
+    public IdentityHubCredentialsVerifier(IdentityHubClient identityHubClient, Monitor monitor, JwtCredentialsVerifier jwtCredentialsVerifier, VerifiableCredentialsJwtUtils JWTPayloadParser) {
+        this.identityHubClient = identityHubClient;
+        this.monitor = monitor;
+        this.jwtCredentialsVerifier = jwtCredentialsVerifier;
+        this.jwtPayloadParser = JWTPayloadParser;
+    }
+
+    /**
+     * Get credentials from the IdentityHub of participant, verifies the credentials, and returns the successfully verified credentials.
+     * To be successfully verified, credentials needs to be signed by the specified issuer in the JWT.
+     *
+     * @param didDocument of a participant. The Did Document should contain an IdentityHub service.
+     * @return VerifiableCredentials.
+     */
+    @Override
+    public Result<Map<String, Object>> getVerifiedCredentials(DidDocument didDocument) {
+        var hubBaseUrl = getIdentityHubBaseUrl(didDocument);
+        if (hubBaseUrl.failed()) return Result.failure(hubBaseUrl.getFailureMessages());
+
+        var jwts = identityHubClient.getVerifiableCredentials(hubBaseUrl.getContent());
+        if (jwts.failed()) {
+            return Result.failure(jwts.getFailureMessages());
+        }
+        var verifiedClaims = jwts.getContent()
+                .stream()
+                .filter(jwt -> jwtCredentialsVerifier.verifyClaims(jwt, didDocument.getId()))
+                .filter(jwtCredentialsVerifier::isSignedByIssuer);
+
+        var credentials = verifiedClaims.map(jwtPayloadParser::extractCredential).collect(Collectors.toList());
```

Comment:
nit: it would be more efficient to use `Collectors.partitioningBy` instead of materializing the stream and iterating twice

Reply:
Good point!